// prism-ai/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/index.ts
var src_exports = {};
__export(src_exports, {
  default: () => pai
});
module.exports = __toCommonJS(src_exports);
var import_ai = require("ai");
var ValidationError = class extends Error {
  constructor(message) {
    super(message);
    this.name = "ValidationError";
  }
};
var BaseClient = class {
  constructor({ api_key, api_url }) {
    this.headers = {
      "Content-Type": "application/json",
      Accept: "application/json",
      Authorization: api_key
    };
    this.api_url = api_url;
  }
};
function safeFetchCall(request) {
  return __async(this, null, function* () {
    const response = yield request;
    if (!response.ok) {
      return Error(response.statusText);
    }
    return yield response.json();
  });
}
var KnowledgeBaseClient = class extends BaseClient {
  create(_0) {
    return __async(this, arguments, function* ({ name }) {
      const methodPath = this.api_url + "/users/knowledge_base/";
      return safeFetchCall(
        fetch(methodPath, {
          method: "POST",
          headers: this.headers,
          body: JSON.stringify({ name })
        })
      );
    });
  }
  get(_0) {
    return __async(this, arguments, function* ({ kb_id }) {
      const methodPath = this.api_url + "/users/knowledge_base/" + kb_id + "/";
      return safeFetchCall(
        fetch(methodPath, {
          method: "GET",
          headers: this.headers
        })
      );
    });
  }
  delete(_0) {
    return __async(this, arguments, function* ({ kb_id }) {
      const methodPath = this.api_url + "/knowledge_base/" + kb_id + "/";
      return safeFetchCall(
        fetch(methodPath, {
          method: "DELETE",
          headers: this.headers
        })
      );
    });
  }
};
var KnowledgeClient = class extends BaseClient {
  create(_0) {
    return __async(this, arguments, function* ({
      method,
      name,
      kb_id,
      url,
      text,
      recursion,
      max_recursion,
      only_base_url
    }) {
      function isValidURL(str) {
        try {
          new URL(str);
          return true;
        } catch (_) {
          return false;
        }
      }
      let appendToMethodPath;
      let methodInput;
      let methodPath;
      const body = {
        name
      };
      switch (method) {
        case "url":
          if (!url || !isValidURL(url)) {
            throw new ValidationError("Invalid or missing URL.");
          }
          methodPath = this.api_url + "/users/knowledge_base/" + kb_id + "/knowledge_from_url/";
          body.url = url;
          if (recursion)
            body.recursion = recursion;
          if (max_recursion)
            body.max_recursion = max_recursion;
          if (only_base_url)
            body.only_base_url = only_base_url;
          break;
        case "text":
          if (!text) {
            throw new ValidationError("Missing text.");
          }
          methodPath = this.api_url + "/users/knowledge_base/" + kb_id + "/knowledge_from_text/";
          body.text = text;
          break;
        default:
          throw new ValidationError(
            "Invalid method. Must be 'url', 'text', or 'path'."
          );
      }
      return safeFetchCall(
        fetch(methodPath, {
          method: "POST",
          headers: this.headers,
          body: JSON.stringify(body)
        })
      );
    });
  }
  get(_0) {
    return __async(this, arguments, function* ({ knowledge_id }) {
      const methodPath = this.api_url + "/knowledge/" + knowledge_id + "/";
      return safeFetchCall(
        fetch(methodPath, {
          method: "GET",
          headers: this.headers
        })
      );
    });
  }
  delete(_0) {
    return __async(this, arguments, function* ({ knowledge_id }) {
      const methodPath = this.api_url + "/knowledge/" + knowledge_id + "/";
      return safeFetchCall(
        fetch(methodPath, {
          method: "DELETE",
          headers: this.headers
        })
      );
    });
  }
};
var ReplyClient = class extends BaseClient {
  create(_0) {
    return __async(this, arguments, function* ({
      prompt,
      conversation_id,
      knowledge_base,
      max_tokens,
      num_results,
      model
    }) {
      const methodPath = this.api_url + "/response/";
      const updatedHeadersForStream = __spreadProps(__spreadValues({}, this.headers), {
        Accept: "application/json"
      });
      const body = {
        user_prompt: prompt
      };
      if (conversation_id)
        body.conversation_id = conversation_id;
      if (knowledge_base)
        body.knowledge_base = knowledge_base;
      if (max_tokens)
        body.max_tokens = max_tokens;
      if (num_results)
        body.num_results = num_results;
      if (model)
        body.model = model;
      const response = yield fetch(methodPath, {
        method: "POST",
        headers: updatedHeadersForStream,
        body: JSON.stringify(body)
      });
      if (!response.body)
        return new Error("No response body.");
      return response;
    });
  }
  stream(_0) {
    return __async(this, arguments, function* ({
      prompt,
      conversation_id,
      knowledge_base,
      max_tokens,
      num_results,
      model
    }) {
      const methodPath = this.api_url + "/response_stream/";
      const updatedHeadersForStream = __spreadProps(__spreadValues({}, this.headers), {
        Accept: "text/response-stream"
      });
      const body = {
        user_prompt: prompt
      };
      if (conversation_id)
        body.conversation_id = conversation_id;
      if (knowledge_base)
        body.knowledge_base = knowledge_base;
      if (max_tokens)
        body.max_tokens = max_tokens;
      if (num_results)
        body.num_results = num_results;
      if (model)
        body.model = model;
      const response = yield fetch(methodPath, {
        method: "POST",
        headers: updatedHeadersForStream,
        body: JSON.stringify(body)
      });
      if (!response.body)
        return new Error("No response body.");
      return new import_ai.StreamingTextResponse(response.body);
    });
  }
};
function pai({
  api_key
}) {
  const api_url = "https://api.prism-ai.ch";
  const knowledgeBaseClient = new KnowledgeBaseClient({ api_key, api_url });
  const knowledgeClient = new KnowledgeClient({ api_key, api_url });
  const replyClient = new ReplyClient({ api_key, api_url });
  return {
    KnowledgeBase: knowledgeBaseClient,
    Knowledge: knowledgeClient,
    Reply: replyClient
  };
}
